core/src/heaviest_subtree_fork_choice.rs

```diff
@@ -80,9 +94,64 @@ struct ForkInfo {
     best_slot: SlotHashKey,
     parent: Option<SlotHashKey>,
     children: Vec<SlotHashKey>,
-    // Whether the fork rooted at this slot is a valid contender
-    // for the best fork
-    is_candidate: bool,
+    // The latest ancestor of this node that has been marked invalid
+    latest_invalid_ancestor: Option<Slot>,
+    is_duplicate_confirmed: bool,
+}
+
+impl ForkInfo {
+    /// Returns if this node has been explicitly marked as a duplicate
+    /// slot
+    fn is_unconfirmed_duplicate(&self, my_slot: Slot) -> bool {
+        self.latest_invalid_ancestor
+            .map(|ancestor| ancestor == my_slot)
+            .unwrap_or(false)
+    }
+
+    /// Returns if the fork rooted at this node is included in fork choice
+    fn is_candidate(&self) -> bool {
+        self.latest_invalid_ancestor.is_none()
+    }
+
+    fn is_duplicate_confirmed(&self) -> bool {
+        self.is_duplicate_confirmed
+    }
+
+    fn set_duplicate_confirmed(&mut self) {
+        self.is_duplicate_confirmed = true;
+        self.latest_invalid_ancestor = None;
+    }
+
+    fn update_with_newly_valid_ancestor(
+        &mut self,
+        my_key: &SlotHashKey,
+        newly_valid_ancestor: Slot,
+    ) {
+        if let Some(latest_invalid_ancestor) = self.latest_invalid_ancestor {
+            if latest_invalid_ancestor <= newly_valid_ancestor {
+                info!("Fork choice for {:?} clearing latest invalid ancestor {:?} because {:?} was duplicate confirmed", my_key, latest_invalid_ancestor, newly_valid_ancestor);
+                self.latest_invalid_ancestor = None;
+            }
+        }
+    }
+
+    fn update_with_newly_invalid_ancestor(
+        &mut self,
+        my_key: &SlotHashKey,
+        newly_invalid_ancestor: Slot,
+    ) {
+        if self
```

Comment:
I think this should also check if the slot is already duplicate confirmed

Reply:
I think we can do one better and add an assert here that `assert!(!fork_info.is_duplicate_confirmed);`: https://github.com/solana-labs/solana/pull/17779/commits/3ad6e3736d8598d7a9d72a4ad5ae990bff07205c#diff-f1f89ae82910cdfebacc27d6ef29a9a55dc89eea426a8e4fcb1586f4ca8da55cR146. 

This is because we should never be marking a duplicate confirmed slot as invalid. There's a test case `test_state_ancestor_confirmed_descendant_duplicate()`https://github.com/solana-labs/solana/blob/7a62cc1e977a6369d9f382ed47c1952e9dbe9546/core/src/cluster_slot_state_verifier.rs#L751 that covers this case

Reply:
Awesome, that was exactly the case I was thinking about